### PR TITLE
Allow timeout parameter at connecting

### PR DESCRIPTION
**Pull request by Elbandi**
php socket_connect doesn't support timeout parameter, the workaround is to use nonblocking mode. The function is already implemented, so i use it.

### CODE REVIEW

**Comment by clue** [comment]
Thanks for this PR! This fixes #7 (and refs #9).

Functionally, this looks good to me :+1: 
However, I'm not sure the API is going to work out once we add additional context parameters, see #9 for details and perhaps also #7 for discussion.

An alternative API could look something like this:

```php
$client = $factory->createClient('tcp://192.168.1.1:1234', array(
    'timeout' => '10'
));
```

What do you think?

**Comment by Elbandi** [comment]
oks. added :)

anyway, the three parameters need to document somehow. but i dont know how (@ foo ?) . :(

File: src/Factory.php [hunk 1]
```diff
@@ -10,18 +10,36 @@ class Factory
...
-            $socket->connect($address);
+            if (isset($context['broadcast']) && $context['broadcast']) {
+                $socket->setOption(SOL_SOCKET, SO_BROADCAST, 1);
+            }
+
+            if (isset($context['bindto'])) {
+                $socket->bind($context['bindto']);
+            }
+
+            if (isset($context['timeout']) && !is_null($context['timeout'])) {
```
Comment (clue) [inline src/Factory.php]: Minor remark: The `is_null()` check is redundant, `isset()` will return `false` on keys that do not exist or have a `null` value.

**Comment by clue** [comment]
Awesome! :) This now fixes #7 and #9.

Afaict the only context option missing is the `blocking` flag, perhaps should also expose this as a separate key?

> […] need to document somehow.

A small docblock (like https://github.com/clue/php-socket-raw/commit/868871dfec979157eaa2c66cdcbd78b453ebd022) is probably sufficient. Either way, I'll probably also update the README before this gets released, so this doesn't necessarily have to be part of this PR.

Again, thanks! :+1:

**Comment by Elbandi** [comment]
Fixed the two things.

My english is not well, so i dont add the docbook help.
I think, the "timeout" parameter need a little doc, others are univoque.

**Comment by clue** [comment]
@Elbandi Thank you very much for working on this PR and your patience!

I've reconsidered this and decided to only add the `$timeout` parameter as you've originally suggested. I've cherry-picked your original commit and merged this via #39. Currently, I don't see an immediate use case for the other parameters, so I've closed #9 accordingly. :shipit:

**Pull request closed without merging**